feat(transformer): support styled components plugin

Dunqing opened the pull request:
close: #11876 

This plugin adds comprehensive support for styled-components with server-side rendering, style minification, and enhanced debugging capabilities. The implementation is ported from the [official Babel plugin](https://github.com/styled-components/babel-plugin-styled-components) to ensure compatibility and feature parity.


## Current Limitations

⚠️ **Import Support**: This plugin currently only supports styled-components imported via ES modules (`import` statements). CommonJS imports using `require("styled-components")` are not yet supported.


## Feature Support

### Options:
 **✅ Fully Supported:**
 - `displayName`: Adds display names for debugging
 - `fileName`: Controls filename prefixing in display names
 - `ssr`: Adds unique component IDs for server-side rendering
 - `transpileTemplateLiterals`: Converts template literals to function calls
 - `minify`: Minifies CSS content in template literals
 - `namespace`: Adds namespace prefixes to component IDs
 - `meaninglessFileNames`: Controls which filenames are considered meaningless

 **⚠️ Partially Supported:**
 - `pure`: Only supports call expressions, not tagged template expressions (bundler limitation)

 **❌ Not Yet Implemented:**
 - `cssProp`: JSX css prop transformation
 - `topLevelImportPaths`: Custom import path handling

## Testing

The test suite is adapted from the [official Babel plugin tests](https://github.com/styled-components/babel-plugin-styled-components) with modifications to integrate with our testing infrastructure. 

**Note**: Some test outputs differ from the original due to our different hashing algorithm for component IDs. This is intentional and follows the same approach used by SWC's implementation.

overlookmotel commented:
I've pushed a commit to enable this plugin in benchmarks, so we can see what impact it has.

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -0,0 +1,1154 @@
+
+    fn enter_call_expression(&mut self, call: &mut CallExpression<'a>, ctx: &mut TraverseCtx<'a>) {
+        if (self.options.display_name || self.options.ssr)
+            // Only transform call expression that is not a part of a member expression
+            // or a callee of another call expression.
+            && !matches!(
+                ctx.parent(),
+                Ancestor::CallExpressionCallee(_) | Ancestor::StaticMemberExpressionObject(_)
+            )
```
Should `ComputedMemberExpression` be transformed?

Dunqing replied:
Aha, it should be, I miss to support `ComputedMemberExpression`. I will do it in the follow-up

overlookmotel reviewed (commented):
I've tried to review this properly, but it's hard because I don't understand what any of it is doing! I know we're keen to get this merged ASAP, but afterwards could you possibly add some comments, and I'll read through it again?

Please feel free to merge in the meantime. I'll do a few follow-up PRs with suggested refactoring and perf, and my comments below could also be addressed in follow-ups.

Another thought: Are we able to disable this plugin on files which aren't JSX/TSX? Or does it need to run on JS files too? The loop through all top-level statements is probably quite costly for long files, and it'd be ideal to skip it if we can.

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -0,0 +1,1154 @@
+        use std::hash::{Hash, Hasher};
+
+        #[inline]
+        fn base36_encode<'a>(mut num: u64, ctx: &TraverseCtx<'a>) -> &'a str {
+            const BASE36_BYTES: &[u8] = b"abcdefghijklmnopqrstuvwxyz0123456789";
+
+            num %= 36_u64.pow(6); // 36^6, to ensure the result is <= 6 characters long.
+
+            let mut result = Vec::with_capacity(6);
+
+            while num != 0 {
+                result.push(BASE36_BYTES[(num % 36) as usize]);
+                num /= 36;
+            }
+
+            ctx.ast.allocator.alloc_str(
+                // SAFETY: the bytes are valid UTF-8 as they are ASCII characters.
+                unsafe { std::str::from_utf8_unchecked(&result) },
+            )
+        }
```
Is there a reason for using base 36 here? Couldn't we use base 64 (A-Z, a-z, 0-9, _, $) instead?

Building the string with `InlineString` would be be more performant than `Vec`. Like in mangler:

https://github.com/oxc-project/oxc/blob/fe843806ece40a19adca3a7c6366abfb6f8662bb/crates/oxc_mangler/src/base54.rs

Dunqing replied:
I guess we can use base64. Using `base36` is to align with its Babel plugin.

https://github.com/styled-components/babel-plugin-styled-components/blob/4e2eb388d9c90f2921c306c760657d059d01a518/src/utils/hash.js#L54-L55. 

Ok, I will follow this kind of optimization.

overlookmotel replied:
It's not a big thing! Was just asking if there was a reason because base36 is unusual.

overlookmotel replied:
@Dunqing I realized now that maybe there *is* a reason. Maybe class names are case insensitive in some browsers?

Dunqing commented on crates/oxc_transformer/src/plugins/mod.rs:
```diff
@@ -0,0 +1,62 @@
+    state::TransformState,
+};
+
+pub struct Plugins<'a, 'ctx> {
+    styled_components: StyledComponents<'a, 'ctx>,
+    options: PluginsOptions,
+}
```
Yes, I thought of this and agree with that second way.

Dunqing commented:
> Not too bad!
> 
> <img alt="bench" width="813" src="https://private-user-images.githubusercontent.com/557937/464400429-133adf92-e625-4df9-b1f1-27e9a1897fc5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIxMDQ5NjEsIm5iZiI6MTc1MjEwNDY2MSwicGF0aCI6Ii81NTc5MzcvNDY0NDAwNDI5LTEzM2FkZjkyLWU2MjUtNGRmOS1iMWYxLTI3ZTlhMTg5N2ZjNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcwOVQyMzQ0MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xN2RkMTdhOWQ4NGNkZWE1NjI1OTMwMWFlN2E5NDRhNTE3ODBjY2IyYTM0NzY5ODUxYWEyZTMwNWUyYmM2NTE0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.i1I1SWbTIanykrNonXFUObP9xbLqEAvZuTOwqDPKHMc">

I guess we don't have a benchmark that can cover all the features that this plugin has, so we don't see much performance change.

Dunqing commented:
Another thought: Are we able to disable this plugin on files which aren't JSX/TSX? Or does it need to run on JS files too? The loop through all top-level statements is probably quite costly for long files, and it'd be ideal to skip it if we can.

It needs to run in all files. One possible case that we need to consider is that they would like to define all of styled components in a utility file (.js or .ts), and use them in the JSX/TSX files.

Dunqing commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -0,0 +1,1154 @@
+    #[serde(default = "default_as_true")]
+    pub ssr: bool,
+
+    /// Transpiles styled-components tagged template literals to a smaller representation
+    /// than what Babel normally creates, helping to reduce bundle size.
+    ///
+    /// Converts `styled.div\`width: 100%;\`` to `styled.div(['width: 100%;'])`, which is
+    /// more compact than the standard Babel template literal transformation.
+    ///
+    /// Default: `true`
+    #[serde(default = "default_as_true")]
+    pub transpile_template_literals: bool,
```
I went through the styled-components documentation and some articles about it, but I don't see any section that describes which way is better for performance. And I don't want to dig into the implementation, so I would prefer to keep this as-is to maintain compatibility.

Dunqing committed:
feat(transformer): support styled components plugin (#12066)

close: #11876

This plugin adds comprehensive support for styled-components with server-side rendering, style minification, and enhanced debugging capabilities. The implementation is ported from the [official Babel plugin](https://github.com/styled-components/babel-plugin-styled-components) to ensure compatibility and feature parity.

## Current Limitations

⚠️ **Import Support**: This plugin currently only supports styled-components imported via ES modules (`import` statements). CommonJS imports using `require("styled-components")` are not yet supported.

## Feature Support

### Options:
 **✅ Fully Supported:**
 - `displayName`: Adds display names for debugging
 - `fileName`: Controls filename prefixing in display names
 - `ssr`: Adds unique component IDs for server-side rendering
 - `transpileTemplateLiterals`: Converts template literals to function calls
 - `minify`: Minifies CSS content in template literals
 - `namespace`: Adds namespace prefixes to component IDs
 - `meaninglessFileNames`: Controls which filenames are considered meaningless

 **⚠️ Partially Supported:**
 - `pure`: Only supports call expressions, not tagged template expressions (bundler limitation)

 **❌ Not Yet Implemented:**
 - `cssProp`: JSX css prop transformation
 - `topLevelImportPaths`: Custom import path handling

## Testing

The test suite is adapted from the [official Babel plugin tests](https://github.com/styled-components/babel-plugin-styled-components) with modifications to integrate with our testing infrastructure.

**Note**: Some test outputs differ from the original due to our different hashing algorithm for component IDs. This is intentional and follows the same approach used by SWC's implementation.

Dunqing commented:
Consider this stack of PR to be huge, so I am merging this stack and following your suggestion and finishing some missing work in the follow-up PRs later.
